README.md
#WideText

v3.1.1

WideText easily makes the font size of one or several lines of text responsive  
It's like Nermal to [the wonderful] [BigText](https://github.com/zachleat/BigText)'s Garfield: younger and lighter.


Bonuses: Just barely more than 1k of vanilla js, runs just once on load (works on window resize events without rerunning)  
Drawback for some: won't work on IE < 8

###[Check out the demo!](http://codepen.io/henry/pen/beBQzJ)

##Useage

**Make text fully responsive simply by wrapping it in `.widetext` and running `wideText()`**  
The **Text** content will be as **Wide** as its container… it'll be **WideText**!  


*Note there will likely be some top and bottom space between the text content and the container, because the container will be sized to contain the largest possible character, from [ascender line to descender line](https://en.wikipedia.org/wiki/Typeface_anatomy#/media/File:Typographia.svg) and including room for [diacritics](https://en.wikipedia.org/wiki/Diacritic). See the [demo](http://codepen.io/henry/pen/beBQzJ) for a visual explanation.*

###Basic example:

Markup:

	<div class="widetext">
	    minimal example
	</div>
	
Script:

	wideText();
	

###Multi-line example
Break the text into several lines by wrapping each line in a `<tspan>`  
The longest line will be the full width of the container,  
and all other lines will get the same font size as the longest.  
**Note: remove whitespace between `tspan`s or you'll get trailing and/or leading space**

Markup:

	<div class="widetext">
        <tspan>multiple</tspan><tspan>lines</tspan>
	</div>

##Options:

The available options are

Option | Default | Attribute for inlining | Possible values
--- | --- | --- | --- |
selector	| `".widetext"` | not available | any valid `"selector"`, as a string
leading* | `.2` | `data-leading` | <p>An em value, whether or not `em` is provided.</p><p>**Inline:** as a string, with or without a unit (e.g. `".2"` or `".2em"`). </p> **Or in js:** string with or without units (e.g. `".2"` or `".2em"`) or a number (e.g. `.2`)
align | null | `data-align` | `"center"` or `"right"` — anything else will align left
row | `"tspan"` | not available | any valid `"element"` or `".class"`, as a string

\* *Note "leading" isn't line height, it's the space between successive lines of text.*


### Custom target

By default **WideText** targets all elements with `.widetext`. To target additional or different elements:

Markup:

    <widetextElement>responsive text</widetextElement>

    <div class="parentOfAnotherWideTextElement">
        <span>responsive text</span>
    </div>

    <yetAnotherWideTextElement>responsive text</yetAnotherWideTextElement>
    
Script:

    wideText("widetextElement"); // shorthand to target an element

    wideText(".parentOfAnotherWideTextElement span"); // shorthand for some other selector

    wideText({selector: "yetAnotherWideTextElement"});


### Custom options

You can pass an array of options like

	.wideText({
		leading: .5,
		align: 'right'
	})

Multiple instances are fine:

	.wideText();
	.wideText({
		selector: '.wideText_alternate',
		row: 'span'
	})

### Data API

Use the data attributes `data-align` and `data-leading` to set the `align` and `leading` options inline. Note that inline options will override options passed in `wideText()`.


**Align multiple lines of text with the `data-align` attribute**  
	
	<div class="widetext" data-align="left">
		<tspan>line 1</tspan><tspan>line 2</tspan>
	</div>	

**Adjust the space between multiple lines with `data-leading`**

	<div class="widetext" data-leading=".2">
		<tspan>line 1</tspan><tspan>line 2</tspan>
	</div>


## Restrictions

The classes `.wideText_svg` and `.wideText_text` are reserved, because after processing

	<div class="myWideText">
		<tspan>line 1</tspan><tspan>line 2</tspan>
	</div>

will be

	<div class="myWideText">
		<svg class="wideText_svg">
			<text class="wideText_text">
				<tspan>line 1</tspan><tspan>line 1</tspan>
			</text>
		</svg>
	</div>

&nbsp;  

--

###Acknowledgments
  
Thanks to [@BigBadaboom](https://github.com/BigBadaboom) and [@dannyjhonston](https://github.com/dannyjhonston) for their help.

CHANGELOG.md
# WideText Changelog

## 3.1.1

*2016-09-27*

Refactor to tighten up - 7% reduction in minified size

## 3.1

*2016-09-27*

Add a major improvement and breaking syntax changes:

Use any element, with the class `.widetext`. Or use a custom selector, matched in the options of `.wideText({options})`.


## 3

*2016-07-09*

v3 will add a major improvement and breaking syntax change:

Use

    <div class="widetext">text</div>

where previously you would have used

    <svg class="widetext"><text>text</text></svg>`

Supports any selector for the target wrapper. Default is `.widetext`.

The resulting markup in the above example will be

    <div class="widetext">
        <svg class="widetext_svg">
            <text class="widetext_text">
                text
            </text>
        </svg>
    </div>



## 2

*2016-07-09*

v2 introduces breaking changes, and introduces a more full-featured tool.

BREAKING CHANGE

WideText must now be run, e.g. with `wideText()`.


NEW FEATURES

To target `svg.customSelector`, run `wideText('.customSelector')` (WideText enforces targetting an svg).

You can pass an array of options like

	.wideText({
		leading: .5,
		align: 'right'
	})

Multiple instances are fine:

	.wideText();
	.wideText({
		selector: '.wideText.spaced-big',
		leading: 3
	})

The available options are

option	| default 		| possible values
---	 	| ---			| ---
selector	| '.wideText' 	| string
leading 	| .2 			| number (e.g. .2) or string (e.g. '.2' or '.2em')
align 	| left align 	| 'center' or 'right'


## 1.0.3

*2016-06-23*

Adds a font size to the svg (6px seems to produce the best end results)


## 1.0.2

*2016-06-23*

Adds Bower registry

Readme updates


## 1.0.1

*2016-06-23*

Makes the `svg` as wide as its relative container


## 1

*2016-06-23*

Initial release

widetext.js
/* WideText

   v3.1.1 - 2016-09-27
   
   https://github.com/olets/WideText
   Copyright (c) 2016 Henry Bley-Vroman (@olets)
   MIT License


 */

function wideText(args) {

    // set options
    var defaults = {
        'selector': '.widetext',
        'leading': .2,
        'align': '',
        'row': 'tspan'
    }
    if (typeof args === "string")
        args = {
            'selector': args
        }
    if (typeof args !== "object")
        args = {};
    for (var option in defaults)
        if (!(option in args))
            args[option] = defaults[option];

        // Determine targets
    var wtElems = document.querySelectorAll(args.selector);

    // For each target
    for (var i = 0; i < wtElems.length; ++i) {

        // Make an <svg> and a <text>, and check for inline options
        var wtElem = wtElems[i],
            svgElem = document.createElementNS("http://www.w3.org/2000/svg", "svg"),
            textElem = document.createElementNS("http://www.w3.org/2000/svg", "text"),
            align = wtElem.getAttribute('data-align') || args.align,
            leadingVal = wtElem.getAttribute('data-leading') || args.leading,
            dyVal = parseFloat(leadingVal, 10) + 1 + "em";

        // Move the target's contents into the <text>
        while (wtElem.firstChild)
            textElem.appendChild(wtElem.firstChild);

        // Apply alignment to the <text>
        if (align == 'center') {
            var anchor = 'middle'
        } else if (align == 'right')
            var anchor = 'end';
        if (anchor) textElem.setAttribute('text-anchor', anchor);
        // And give it the WideText class
        textElem.classList.add("wideText_text");

        // Put each row in a <tspan>
        var rowElems = textElem.querySelectorAll(args.row);
        for (j = 0; j < rowElems.length; j++) {
            var rowElem = rowElems[j],
                tspanElem = document.createElementNS("http://www.w3.org/2000/svg", "tspan");
            while (rowElem.firstChild)
                tspanElem.appendChild(rowElem.firstChild);

            // And place the <tspan>
            tspanElem.setAttribute('x', '0');
            tspanElem.setAttribute('dy', dyVal);

            // And put the <tspan>s in the <text>
            textElem.replaceChild(tspanElem, rowElem);
        }

        // Put the <text> in the <svg>
        svgElem.appendChild(textElem);

        // And apply default styles
        svgElem.style.width = "100%";
        svgElem.style.display = "block";
        svgElem.style["font-size"] = "6px";

        // And give it the WideText class
        svgElem.classList.add("wideText_svg");

        // Put the <svg> in the target
        wtElem.appendChild(svgElem);

        // Size the <svg> to fit the <text>
        var textBBox = textElem.getBBox();
        svgElem.setAttribute("viewBox", [textBBox.x, textBBox.y, textBBox.width, textBBox.height].join(' '));
    }
}
